16.0 mig commown factor out transaction label

Honeyxilia opened the pull request.

Honeyxilia committed:
Create contract_transaction_label for custom invoice labels

Honeyxilia committed:
[test] Replaced ref method in test

BaseCase.ref is an instance method, which means it can't be used in the class method setUpClass.
As such, we're using the environnement ref method.

Honeyxilia committed:
[test] Changed payment.token record creation

Honeyxilia committed:
Simplified transaction_label field insertion in views

The main group_auto_pay group we're placing the label field relative to
is already filtered out with the is_auto_pay boolean value.
As such, we can directly insert the field into the group_auto_pay block
without adding the condition.

Honeyxilia committed:
Adapted to account.move field changes

Honeyxilia committed:
[test] Added Slimpay provider test enabling

In the payment module _send_payment_request overload, a UserError is raised if the provider is disabled, which by default, the Slimpay provider is.
As such, we enable the Slimpay provider as test, to be able to perform the payment request.

Honeyxilia committed:
[test] Added date_start overwrite to match recurring_next_date

When computing next_period_date_end, it computes next_period_date_start,
which is based on the contract line date_start field if last_date_invoiced is not set.
As such, it calculates not from the wanted recurring_next_date, but from date_start.

To avoid that, we simply set the contract date_start to the recurring_next_date value, to fill the data
to match the wanted period.

Honeyxilia committed:
[test] Added Slimpay provider assignment to payment method when account_payment is installed

When the account_payment module is installed (alongside the sale module), a new field
is added to account.payment.method, payment_provider_id.
Then, when creating transactions, any payment methods which don't match the provider assigned in the transaction
are filtered out (see odoo/addons/account_payment/models/payment_transaction.py:L139).

As such, when the module is installed, we assign the Slimpay provider to the journal payment methods
to be able to pass the tests.

Honeyxilia committed:
[coverage] Added a test for contracts with no transaction label

fcayre commented on account_payment_slimpay_transaction_label/views/contract_contract.xml:
```diff
@@ -5,15 +5,10 @@
-                >
-          <group>
+      <xpath expr="//group[@name='group_auto_pay']" position="inside">
+      <group name="group_transaction">
```
Le <group> a aussi un rôle graphique donc il n'est pas équivalent de mettre un groupe dans un autre ou après un autre. Est-ce que tu as pris le temps de vérifier l'impact visuel ?

Idem dans le même commit dans contract_template.xml.

Par ailleurs dans github il semble y avoir un souci d'indentation ici (tags xpath et group alignés), c'est bizarre.

Honeyxilia replied:
A vrai dire, j'avais justement fait cette modification car les balises `group` imbriquées rendaient mal visuellement sur l'interface

fcayre commented on account_payment_slimpay_transaction_label/tests/test_contract.py:
```diff
@@ -65,3 +65,21 @@ def test_custom_payin_label(self):
+            self.contract.with_context(
+                slimpay_payin_label=slimpay_payin_label
+            ).recurring_create_invoice()
+            label = pay.call_args[0][-1]
```
Les lignes 84 et 85 devraient être en dehors du `with patch`, afin de rendre clair l'objectif de ce context manager : faire passer `recurring_create_invoice(...)`

fcayre reviewed (changes requested):
Techniquement seuls les tests dépendent de Slimpay. J'aurais donc plutôt : 
1. sorti un module nommé contract_transaction_label qui ne dépend pas de Slimpay mais fixe une "API" à respecter par les payment.provider qui veulent la supporter : utiliser une variable de contexte, nommée par exemple "payment_transaction_label" (attention léger impact dans payment_slimpay_issue qui utilise aussi slimpay_payin_label)
2. dans les tests, uniquement vérifier que cette variable de contexte est disponible lors de la création du payment, avec n'importe quel provider de odoo.core
3. idéalement, créé un autre module, account_payment_slimpay_transaction_label qui s'auto-installe quand ses 2 dépendances (contract_transaction_label et account_payment_slimpay) le sont et qui surcharge la méthode _label de PaymentTransation en modifiant son implémentation dans account_payment_slimpay.

Qu'en dis-tu ?

Honeyxilia commented:
Je suis d'accord avec la création d'un module intermédiaire générique - cependant, comment puis-je vérifier que la valeur soit établie dans le contexte dans `_pay_invoice` ?

Honeyxilia committed:
Created new module account_payment_slimpay_transaction_label

Honeyxilia committed:
Modified _label method implementation to better fit the dependency chain

Since the label context value is mainly used in the contract_transaction_label, we use this new module linking Slimpay and the transaction label,
to try to fetch the label in this module, rather than account_payment_slimpay.

Honeyxilia committed:
Move code out of commown into contract_transaction_label

Honeyxilia committed:
[test] Moved out contract tests related to Slimpay custom label

Honeyxilia committed:
[test] Adapted to account.move field change

Honeyxilia committed:
Renamed env. label context variable name to be more generic

Honeyxilia committed:
Create new test module test_contract_slimpay_transaction_label

Honeyxilia committed:
[test] Moved useful methods in dedicated mixin class

Honeyxilia committed:
[test] Added tests for _label method overload

Honeyxilia committed:
[test] Added generic tests for computing the transaction label from a contract

These tests were inspired by the commown tests, without using Slimpay as the provider,
but a base provider from the `payment` module.

Honeyxilia committed:
[test] Removed tabulations to better encapsulate the operation concerned by the patch

fcayre commented on test_contract_slimpay_transaction_label/tests/test_contract.py:
```diff
@@ -0,0 +1,59 @@
+            {
+                "transaction_label": "Invoice #START# - #END# (#INV#)",
+            }
+        )
```
Ici on peut privilégier la forme plus courte / lisible : 
`self.contract.transaction_label = "Invoice #START# - #END# (#INV#)"`

Honeyxilia replied:
j'ai appliqué cette remarque c:

fcayre reviewed (approved):
Une mini-remarque de pure forme absolument pas indispensable. Tu peux fusionner ! \o/

Honeyxilia committed:
[test] Inlined record field assignement

Honeyxilia commented:
/ocabot merge patch
